Source/WebCore/platform/graphics/ImageFrameProvider.cpp

```diff
@@ -0,0 +1,392 @@
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ImageFrameProvider.h"
+
+#include "BitmapImageSource.h"
+#include "GraphicsContext.h"
+#include "ImageBuffer.h"
+#include "ImageDecoder.h"
+#include "ImageFrame.h"
+#include "Logging.h"
+
+namespace WebCore {
+
+Ref<ImageFrameProvider> ImageFrameProvider::create(BitmapImageSource& source, Ref<ImageDecoder>&& decoder)
+{
+    return adoptRef(*new ImageFrameProvider(source, WTFMove(decoder)));
+}
+
+ImageFrameProvider::ImageFrameProvider(BitmapImageSource& source, Ref<ImageDecoder>&& decoder)
+    : m_source(source)
+    , m_decoder(WTFMove(decoder))
+{
+}
+
+ImageFrameProvider::~ImageFrameProvider()
+{
+    stopDecodingWorkQueue();
+}
+
+Ref<ImageFrameWorkQueue> ImageFrameProvider::protectedWorkQueue() const
+{
+    if (!m_workQueue)
+        m_workQueue = ImageFrameWorkQueue::create(const_cast<ImageFrameProvider&>(*this));
+    return Ref { *m_workQueue };
+}
+
+void ImageFrameProvider::encodedDataStatusChanged(EncodedDataStatus status)
+{
+    if (status < EncodedDataStatus::SizeAvailable)
+        return;
+
+    m_frames.resizeToFit(protectedDecoder()->frameCount());
+}
+
+void ImageFrameProvider::destroyDecodedData(bool destroyAll)
+{
+    LOG(Images, "ImageFrameProvider::%s - %p - url: %s. Decoded data with destroyAll = %d will be destroyed.", __FUNCTION__, this, sourceUTF8(), destroyAll);
+
+    for (unsigned index = 0; index < frameCount(); ++index) {
+        // Don't destroy a NativeImage betore it is drawn at least once.
+        if (isPendingDrawingAtIndex(index))
+            continue;
+
+        if (!destroyAll && !protectedSource()->canDestroyNativeImageAtIndex(index))
+            continue;
+
+        destroyNativeImageAtIndex(index);
+    }
+
+    decodePropertiesChanged(0);
+}
+
+void ImageFrameProvider::decodedSizeIncreased(unsigned decodedSize)
+{
+    if (!decodedSize)
+        return;
+
+    m_decodedSize += decodedSize;
+
+    protectedSource()->decodedSizeChanged(decodedSize);
+    decodePropertiesChanged(0);
+}
+
+void ImageFrameProvider::decodedSizeDecreased(unsigned decodedSize)
+{
+    if (!decodedSize)
+        return;
+
+    ASSERT(m_decodedSize >= decodedSize);
+    m_decodedSize -= decodedSize;
+
+    protectedSource()->decodedSizeChanged(-static_cast<long long>(decodedSize));
+}
+
+void ImageFrameProvider::decodePropertiesChanged(unsigned decodedPropertiesSize)
+{
+    if (m_decodedSize)
+        return;
+
+    long long decodedSize = static_cast<long long>(decodedPropertiesSize) - m_decodedPropertiesSize;
+    m_decodedPropertiesSize = decodedPropertiesSize;
+    protectedSource()->decodedSizeChanged(decodedSize);
```

Comment:
I don't think we'll ever get a chance to subtract `m_decodedPropertiesSize` from the decoded size, will we? Because we early exit from this function if `m_decodedSize` is non-null, and in `decodedSizeIncreased` (which should be when we first decode some bytes from the image data itself), we call `decodePropertiesChanged(0)` after increasing `m_decodedSize`.

Reply:
Yes you are right. It seems the original code has some flaws and my implementation made it even worse. I think the rules for setting `m_decodedPropertiesSize` and `m_decodedSize` should be the following:

1. In `BitmapImageMetadata::imageMetadata()` and after decoding the image meta data, `m_decodedPropertiesSize` will be set to `ImageDecoder::bytesDecodedToDetermineProperties()` only if `m_decodedSize` is zero. This will be true only if we have not decoded any frames. MemoryCache will be increased by `m_decodedPropertiesSize`.
2. When decoding an image frame, `m_decodedSize` will be increased by `decodedSize`, `decodedSize` will be subtracted  by `m_decodedPropertiesSize`, `m_decodedPropertiesSize` will be set to zero. MemoryCache will be increased by `decodedSize`.
3. In ImageFrameProvider::destroyDecodedData() `m_decodedSize` will be decreased by `decodedSize`. MemoryCache will be decreased by `decodedSize`. `m_decodedPropertiesSize` will not be changed.